src/dispatcher.cpp

```diff
@@ -88,28 +95,21 @@ void dispatcher::stop()
 }
 
 
-// Return when all items in the queue are finished (within a timeout).
+// Return when all current items in the queue are finished (within a timeout).
 // If additional items are added while we're waiting, those will not be waited on!
+// Returns false if a timeout occurred before we were done
 //
 bool dispatcher::flush()
 {
-    std::mutex m;
-    std::condition_variable cv;
-    bool invoked = false;
-    auto wait_sucess = std::make_shared<std::atomic_bool>(true);
-    invoke([&, wait_sucess](cancellable_timer t)
-    {
-        ///TODO: use _queue to flush, and implement properly
-        if (_was_stopped || !(*wait_sucess))
-            return;
-
-        {
-            std::lock_guard<std::mutex> locker(m);
-            invoked = true;
-        }
-        cv.notify_one();
-    });
-    std::unique_lock<std::mutex> locker(m);
-    *wait_sucess = cv.wait_for(locker, std::chrono::seconds(10), [&]() { return invoked || _was_stopped; });
-    return *wait_sucess;
+    if( _was_stopped )
+        return true;  // Nothing to do - so success (no timeout)
+
+    utilities::time::waiting_on< bool > invoked( false );
+    invoke( [invoked = invoked.in_thread()]( cancellable_timer ) {
+        invoked.signal( true );
+    } );
+    invoked.wait_until( std::chrono::seconds( 10 ), [&]() {
+        return invoked || _was_stopped;
```

Comment:
In case `_was_stopped` is toggled on during `flush` the correctness of `flush` should be verified

Reply:
What do you mean by 'the correctness of flush'?

It states:
```
// Returns false if a timeout occurred before we were done
```
Are you saying you want this changed?

Reply:
I.e., if we're waiting for a flush and someone else stops the dispatcher, this function will return immediately, as if the timeout occurred.

Reply:
Correct, so if during flash `_was_stopped` is activated, the flush operation is interrupted and terminates abnormally.
But I think one needs to wait for `flush` to accomplish properly for graceful transition

Reply:
I don't understand.

What's a graceful transition? We reached our timeout, the invoke never happened. What else is there to wait on?